### PR TITLE
Implement send and receive text and file

### PR DESCRIPTION
**Pull request by bryanchriswhite**
This PR implements a dart, object-oriented API that wraps the cgo API that is exported by `wormhole-william` in accordance with https://github.com/LeastAuthority/dart_wormhole_william/issues/7.

Depends on https://github.com/LeastAuthority/wormhole-william/pull/26

Additionally, #3 and #4 are implemented in this PR.

---

## Code Review Checklist (to be filled out by reviewer)

- [x] Description accurately reflects what changes are being made.
- [x] Description explains why the changes are being made (or references an issue containing one).
- [ ] The PR appropriately sized.
- [ ] New code has enough tests. 
- [ ] New code has enough documentation to answer "how do I use it?" and "what does it do?".
- [x] Existing documentation is up-to-date if impacted.

### CODE REVIEW

File: android/CMakeLists.txt [hunk 1]
```diff
@@ -23,9 +27,8 @@ else()
   set(GOARCH "please-fail :/")
```
Comment (ewanas) [inline android/CMakeLists.txt]: Definitely should use a fail command like the one used for `FLUTTER_HOME`

File: android/CMakeLists.txt [hunk 2]
```diff
@@ -23,9 +27,8 @@ else()
...
 endif()
 
-set(WORMHOLE_WILLIAM_GO_DIR ${CMAKE_SOURCE_DIR}/../wormhole-william)
-
-set(CC_FOR_TARGET ${ANDROID_TOOLCHAIN_ROOT}/bin/${CMAKE_C_COMPILER_TARGET}-clang)
+set(CC_FOR_TARGET
+    ${ANDROID_TOOLCHAIN_ROOT}/bin/${CMAKE_C_COMPILER_TARGET}-clang)
 
 # Fix the clang triplet used as parameter for -target flag. compiler with -none-
 # are not in the ndk
```
Comment (ewanas) [inline android/CMakeLists.txt:30]: Figure out if we could clean these text replacements on `CC_FOR_TARGET`

File: android/CMakeLists.txt [hunk 3]
```diff
@@ -35,6 +38,12 @@ string(REPLACE "-none" "" CC_FOR_TARGET ${CC_FOR_TARGET})
 # architecture than the one that comes with the ndk armv7 vs armv7a
 string(REPLACE "armv7-" "armv7a-" CC_FOR_TARGET ${CC_FOR_TARGET})
 
-add_library(${PLUGIN_NAME} SHARED "src/main/c/dummy.c")
+add_library(${PLUGIN_LIB} SHARED "src/main/c/dummy.c")
```
Comment (ewanas) [inline android/CMakeLists.txt:37]: See if we can avoid having a dummy C file

File: android/CMakeLists.txt [hunk 4]
```diff
@@ -35,6 +38,12 @@ string(REPLACE "-none" "" CC_FOR_TARGET ${CC_FOR_TARGET})
...
+include(${DART_WORMHOLE_WILLIAM_C_DIR}/../../CMakeLists.txt)
+include(${DART_WORMHOLE_WILLIAM_C_DIR}/CMakeLists.txt)
+
+target_link_libraries(${PLUGIN_LIB} PRIVATE bindings)
```
Comment (ewanas) [inline android/CMakeLists.txt:41]: Cleanup here instead of two commands

File: android/CMakeLists.txt [hunk 5]
```diff
@@ -35,6 +38,12 @@ string(REPLACE "-none" "" CC_FOR_TARGET ${CC_FOR_TARGET})
...
+target_link_libraries(${PLUGIN_LIB} PRIVATE wormhole_william)
 
-include(../CMakeLists.txt)
+set(dart_wormhole_william_bundled_libraries "" $<TARGET_FILE:wormhole_william>
```
Comment (ewanas) [inline android/CMakeLists.txt]: Double check about that `""`

File: lib/client/client.dart [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    Pointer<Pointer<Utf8>> codeOut = calloc();
+
+    final rxPort = ReceivePort()
+      ..listen((dynamic errCode) {
+        if (errCode is int) {
+          // TODO: Create exception implementation(s).
+          final exception =
+              Exception('Failed to send text. Error code: $errCode');
+          done.completeError(exception);
+          throw exception;
```
Comment (bryanchriswhite) [inline lib/client/client.dart]: Does this `throw` make sense?

File: lib/client/client.dart [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+      });
+
+    final errCode = _native.clientSendText(
+        msg.toNativeUtf8(), codeOut, rxPort.sendPort.nativePort);
+
+    if (errCode != 0) {
+      // TODO: Create exception implementation(s).
+      calloc.free(codeOut);
+      final exception = Exception('Failed to send text. Error code: $errCode');
+      done.completeError(exception);
```
Comment (bryanchriswhite) [inline lib/client/client.dart]: Don't think this makes sense.

File: lib/client/client.dart [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+class Client {
+  late NativeClient _native;
+  late Config? _config;
+
+  Client({Config? config}) {
+    _native = NativeClient(config: config);
+    _config = config;
+  }
+
+  Future<SendResult> sendText(String msg) {
```
Comment (bryanchriswhite) [inline lib/client/client.dart]: Make this actually async.

File: lib/client/client.dart [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    final done = Completer<String>();
+
+    final rxPort = ReceivePort()
+      ..listen((dynamic response) {
+        if (response is int) {
+          // TODO: Create exception implementation(s).
+          final exception =
+              Exception('Failed while sending text. Error code: $response');
+          done.completeError(exception);
+          throw exception;
```
Comment (bryanchriswhite) [inline lib/client/client.dart]:
```suggestion
```

File: lib/client/client.dart [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    final Pointer<Utf8> code = codeOut.value;
+    calloc.free(codeOut);
+
+    final result = SendResult(code.toDartString(), done.future);
+    return Future.value(result);
+  }
+
+  Future<String> recvText(String code) {
```
Comment (bryanchriswhite) [inline lib/client/client.dart]: Make actually async

File: lib/client/client.dart [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+      throw Exception('Failed to send text. Error code: $errCode');
+    }
+
+    return done.future;
+  }
+
+  Future<SendResult> sendFile(File file) async {
+    final fileName = path.basename(file.path);
+    final length = await file.length();
+    print("File length was: $length");
```
Comment (bryanchriswhite) [inline lib/client/client.dart]:
```suggestion
```

File: lib/client/client.dart [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    final done = Completer<void>();
+
+    Pointer<Pointer<Utf8>> codeOut = calloc();
```
Comment (bryanchriswhite) [inline lib/client/client.dart:123]: We should look into this, what is `calloc` doing exactly?

File: lib/client/client.dart [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    // TODO: figure out if we can avoid copying data.
+    // e.g. ByteData | ByteBuffer | Uint8Array (?)
+    final Pointer<Uint8> nativeBytes =
+        malloc(length); // Allocator<Uint8>.allocate(length);
+    // Uint8Array(length);
+
+    final rxPort = ReceivePort()
+      ..listen((dynamic errCode) {
+        print("Got ${errCode}");
```
Comment (bryanchriswhite) [inline lib/client/client.dart]:
```suggestion
```

File: lib/client/client.dart [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        if (errCode is int) {
+          // TODO: Create exception implementation(s).
+          throw Exception(
```
Comment (bryanchriswhite) [inline lib/client/client.dart]: This should call `completeError` on the completer and not throw an exception.

File: lib/client/exceptions.dart [hunk 15]
```diff
@@ -0,0 +1,3 @@
+enum ErrorCodes {
```
Comment (bryanchriswhite) [inline lib/client/exceptions.dart:1]: I don't think we use this. Clean up.

File: lib/client/native_client.dart [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+typedef InitDartApiNative = IntPtr Function(Pointer<Void>);
+typedef InitDartApi = int Function(Pointer<Void>);
+
+typedef NewClientNative = IntPtr Function(
+    Pointer<Utf8> appId,
+    Pointer<Utf8> rendezvousUrl,
+    Pointer<Utf8> transitRelayUrl,
+    Int32 passPhraseComponentLength);
+typedef NewClient = int Function(
+    Pointer<Utf8> appId,
```
Comment (bryanchriswhite) [inline lib/client/native_client.dart:17]: Can / should these be `String`?

File: lib/client/native_client.dart [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+typedef ClientRecvText = int Function(
+    int goClientId, Pointer<Utf8> code, int callbackPortId);
+
+typedef ClientRecvFileNative = Int32 Function(
+    Uint32 goClientId, Pointer<Utf8> code, Int32 callbackPortId);
+
+typedef ClientRecvFile = int Function(
+    int goClientId, Pointer<Utf8> code, int callbackPortId);
+
+const DEFAULT_APP_ID = "lothar.com/wormhole/text-or-file-xfer";
```
Comment (bryanchriswhite) [inline lib/client/native_client.dart]: Allow override via environment variables or config file at build-time, rely on go for defaults.

File: lib/client/native_client.dart [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+class NativeClient {
+  late final DynamicLibrary _wormholeWilliamLib;
+  late final DynamicLibrary _asyncCallbackLib;
+  late final _goClientId;
+
+  String _appId = DEFAULT_APP_ID;
+  String _rendezvousUrl = DEFAULT_RENDEZVOUS_URL;
+  String _transitRelayUrl = DEFAULT_TRANSIT_RELAY_URL;
+  int _passPhraseComponentLength = DEFAULT_PASSPHRASE_COMPONENT_LENGTH;
```
Comment (bryanchriswhite) [inline lib/client/native_client.dart]: Simplify. We can use a new / default config in the constructor.

File: lib/client/native_context.dart [hunk 19]
```diff
@@ -0,0 +1,6 @@
+import 'dart:ffi';
+
+class NativeContext extends Struct {
```
Comment (bryanchriswhite) [inline lib/client/native_context.dart]: Double-check if we're using this.

File: lib/src/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,22 @@
+if(NOT DEFINED ENV{FLUTTER_HOME})
+  message(FATAL_ERROR "Flutter home must be defined")
+endif()
+
+add_library(
+  bindings SHARED ${DART_WORMHOLE_WILLIAM_C_DIR}/async_callback.c
+                  $ENV{FLUTTER_HOME}/bin/cache/dart-sdk/include/dart_api_dl.c)
+target_include_directories(
+  bindings PRIVATE $ENV{FLUTTER_HOME}/bin/cache/dart-sdk/include)
+target_include_directories(bindings PRIVATE ${flutter_INCLUDE_DIRECTORIES})
```
Comment (bryanchriswhite) [inline lib/src/CMakeLists.txt]: Double-check if needed or redundant.

File: lib/src/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+target_include_directories(bindings PRIVATE ${WORMHOLE_WILLIAM_INCLUDE_DIR})
+target_include_directories(bindings PRIVATE ${WORMHOLE_WILLIAM_C_DIR})
+target_link_libraries(bindings PRIVATE $<TARGET_FILE:wormhole_william>)
+
+set_target_properties(
+  bindings
+  PROPERTIES PUBLIC_HEADER bindings.h
+             PUBLIC_HEADER include/dart_version.h
```
Comment (bryanchriswhite) [inline lib/src/CMakeLists.txt]: Do these need to be `PUBLIC_HEADER`s?

File: lib/src/async_callback.c [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <dart_api.h>
+#include <dart_api_dl.h>
+
+#include "libwormhole_william.h"
+// TODO define this in the build
+#define DEBUG
```
Comment (bryanchriswhite) [inline lib/src/async_callback.c]: Move to CMakeLists

File: lib/client/client.dart (hunk 12 above)
Comment (ewanas) [reply lib/client/client.dart:123]: To reiterate our joint digging for this one, `calloc` is an `Allocator` and allocators are callable due to an extension function defined on `Allocator` that takes an optional count(default is 1).

File: android/CMakeLists.txt (hunk 2 above)
Comment (ewanas) [reply android/CMakeLists.txt:30]: Tried removing either one and the issue remains; these replacements seem needed for now

File: android/CMakeLists.txt (hunk 5 above)
Comment (ewanas) [reply android/CMakeLists.txt]: Removed

File: lib/src/CMakeLists.txt (hunk 20 above)
Comment (ewanas) [reply lib/src/CMakeLists.txt]: Removed. The preceding include directory is sufficient.

File: lib/src/CMakeLists.txt (hunk 21 above)
Comment (ewanas) [reply lib/src/CMakeLists.txt]: Removed. They were superfluous

File: android/CMakeLists.txt (hunk 3 above)
Comment (ewanas) [reply android/CMakeLists.txt:37]: Seems like we're stuck with this hack for now. The issue is that linux(and maybe other platforms) have code within that library, but for Android we didn't have such a need.

File: lib/client/native_client.dart (hunk 18 above)
Comment (ewanas) [reply lib/client/native_client.dart]: Did a bit of refactoring to clean this up. `NativeClient` now either uses the provided `Config` or creates a `Config` with no explicit parameters, thereby using the defaults.

File: android/CMakeLists.txt (hunk 1 above)
Comment (ewanas) [reply android/CMakeLists.txt]: Added explicit failure here

File: lib/client/client.dart (hunk 7 above)
Comment (ewanas) [reply lib/client/client.dart]: Removed

File: lib/client/client.dart (hunk 6 above)
Comment (ewanas) [reply lib/client/client.dart]: Removed

File: lib/src/async_callback.c (hunk 22 above)
Comment (ewanas) [reply lib/src/async_callback.c]: Implementation added to the bindings CMakeLists.txt file.

**Review by shareef-dweikat — APPROVED** [review]

**Pull request merged**